### Constructing a Decision Tree 


The problem of constructing a decision tree can be expressed recursively. First, select an attribute to place at the root node and make one branch for each possible value. This splits up the example set into subsets, one for every value of the attribute. 

Now the process can be repeated recursively for each branch, using only those instances that actually reach the branch. If at any time all instances at a node have the same classification, stop developing that part of the tree.

The only thing left to decide is how to determine which attribute to split on,
given a set of examples with different classes.Consider the weather data. 

![](img/weather-data.png)

There are four possibilities for each split.

![](img/tree-top.png)

Which is the best choice? The number of yes and no classes are shown at the leaves. Any leaf with only one class—yes or no—will not have to be split further, and the recursive process down that branch will terminate.

Because we seek *small trees*, we would like this to happen as soon as possible. If we had a measure of the **purity** of each node, we could choose the attribute that produces the purest daughter nodes. A pure node would be one where there are only examples belonging to only a single class. 

The measure of purity that we will use is called the information entropy and is measured in units called bits. How do we calculate this???

We calculate it based on the number of yes and no classes at the node. When evaluating the first tree (outlook tree), the numbers of yes and no classes at the leaf nodes are [2,3], [4,0], and [3,2].

We calculate the information gain for each attribute and split on the one that gains the most information.


The information we calculate has these properties:
1. When the number of either yes’s or no’s is zero, the information is zero.
2. When the number of yes’s and no’s is equal, the information reaches a maximum.

Information value or entropy is calculated as 

$
entropy(p_1,p_2, .. p_n) = -p_1 log_2 p_1 - p_2 log_2 p_2,  ... - p_n log_2 p_n
$

The Outlook attribute has three values, sunny, overcast, rainy. The sunny value yields 2 positive and 3 negative classification result if we used this attribute for classification. 



Reference: 
- *Witten & Frank. 2005. Data Mining Practical Machine Learning Tools and Techniques, 2nd Ed. Morgan Kaufman*
- *Quinlan, J. R. 1986. Induction of Decision Trees. Mach. Learn. 1, 1 (Mar. 1986), 81–106*

In [2]:
from scipy.stats import entropy

sunny_info=entropy([2/5,3/5], base=2)
print('Entropy (sunny):', sunny_info)

Entropy (sunny): 0.9709505944546688


Verify that calculated information values of these nodes for Outlook are:

sunny_info([2,3])=0.971 bits
overcast_info([4,0])=0.0 bits
rainy_info([3,2)]=0.971 bits

After calculating the entropy for each attribute value, we obtain the average entropy for the node Outlook. Outlook([2+,3-], [4+,0-], [3+,2-]) is given as below:

In [6]:
Outlook=5./14. * 0.971 + 4./14. * 0 +  5./14. * 0.971
Outlook

0.6935714285714286

We use the **Gain** to decide which node is used to split the tree. To do that we need to calculate the **Gain** for all possible splits (all attributes)

We calculate the *gain differences* before and after the split.  The entropy of the problem before any tree is split is the entropy of 9+ and 5-, corresponding to:


In [7]:
tree=entropy([9./14.,5./14.], base=2)
tree

0.940285958670631

The change or the information gain obtained after the split at Outlook is:

In [8]:
gain_outlook=tree-Outlook
print('Gain(outlook):', gain_outlook)

Gain(outlook): 0.2467145300992024


Gain(outlook) can be interpreted as the informational value of creating a branch on the outlook attribute.

The gain for each attribute split can be calculated as:

- gain(outlook) = 0.247 bits
- gain(temperature) = 0.029 bits
- gain(humidity) = 0.152 bits
- gain(windy) = 0.048 bits

Therefore, we select outlook as the splitting attribute at the root of the tree. Hopefully this accords with your intuition as the best one to select. It is the only choice for which one daughter node is completely pure, and this gives it a considerable advantage over the other attributes. Humidity is the next best choice because it produces a larger daughter node that is almost completely pure.

### Question
1. Verify the gain for temperature, humidity and windy


temp

In [58]:
hot_info=entropy([2/4,2/4], base=2)
mild_info=entropy([4/6,2/6], base=2)
cool_info=entropy([3/4,1/4], base=2)

print('Entropy (hot):', hot_info)
print('Entropy (mild):', mild_info)
print('Entropy (cool):', cool_info)

Entropy (hot): 1.0
Entropy (mild): 0.9182958340544894
Entropy (cool): 0.8112781244591328


In [70]:
temp=4/14 * 1 +  6/14 * 0.9182958340544894 +  4/14 * 0.8112781244591328
temp

0.9110633930116763

In [71]:
temp_tree=entropy([9./14.,5./14.], base=2)
temp_tree

0.940285958670631

In [72]:
gain_temp=temp_tree-temp
print('Gain(temp):', gain_temp)

Gain(temp): 0.029222565658954758


humidity

In [54]:
high_info=entropy([3/7,4/7], base=2)
normal_info=entropy([6/7,1/7], base=2)
print('Entropy (high):', high_info)
print('Entropy (normal):', normal_info)

Entropy (high): 0.9852281360342515
Entropy (normal): 0.5916727785823274


In [77]:
humidity=7./14. * 0.9852281360342515  +  7./14. * 0.5916727785823274
humidity

0.7884504573082894

In [78]:
humidity_tree=entropy([9./14.,5./14.], base=2)
humidity_tree

0.940285958670631

In [80]:
gain_humidity=humidity_tree-humidity
print('Gain(humidity):', gain_humidity)

Gain(humidity): 0.15183550136234159


windy

In [73]:
false_info=entropy([6/8,2/8], base=2)
true_info=entropy([3/6,3/6], base=2)
print('Entropy (false):', false_info)
print('Entropy (true):', true_info)

Entropy (false): 0.8112781244591328
Entropy (true): 1.0


In [74]:
windy=8./14. * 0.8112781244591328  +  6./14. * 1
windy

0.8921589282623617

In [75]:
windy_tree=entropy([9./14.,5./14.], base=2)
windy_tree

0.940285958670631

In [76]:
gain_windy=windy_tree-windy
print('Gain(windy):', gain_windy)

Gain(windy): 0.04812703040826938


Decision tree algorithms are based on a divide-and-conquer approach to the classification problem. They work top-down, seeking at each stage an attribute to split on that best separates the classes, and then recursively processing the subproblems that result from the split. This strategy generates a decision tree

![](img/tree-2.png)

Then we continue, recursively. The figure shows the possibilities for a further
branch at the node reached when outlook is sunny. Clearly, a further split on
outlook will produce nothing new, so we only consider the other three attributes.

The information gain for each turns out to be :

- gain(temperature)=0.571 bits
- gain(humidity)=0.971 bits
- gain(windy)=0.020 bits

so we select humidity as the splitting attribute at this point. There is no need to split these nodes any further, so this branch is finished.

Ideally, the process terminates when all leaf nodes are pure, that is, when they contain instances that all have the same classification.

Finally we may have a decision tree as below:

![](img/tree-3.png)

### Question
1. Verify the information gain for temperature, humidity and windy above

 temperature

In [81]:
hot_info=entropy([0/2,2/2], base=2)
mild_info=entropy([1/2,1/2], base=2)
cool_info=entropy([0/1,1/1],base=2)

print('Entropy (hot):', hot_info)
print('Entropy (mild):', mild_info)
print('Entropy (cool):', cool_info)

Entropy (hot): 0.0
Entropy (mild): 1.0
Entropy (cool): 0.0


In [82]:
temp=2/5 * 0 +  2/5 * 1 +  1/5 * 0
temp

0.4

In [83]:
temp_tree=entropy([2./5.,3./5.], base=2)
temp_tree

0.9709505944546688

In [84]:
gain_temp=temp_tree-temp
print('Gain(temp):', gain_temp)

Gain(temp): 0.5709505944546688


humidity

In [85]:
hith_info=entropy([0/3,3/3], base=2)
normal_info=entropy([2/2,0/2], base=2)

print('Entropy (high):', hith_info)
print('Entropy (mild):', normal_info)

Entropy (hot): 0.0
Entropy (mild): 0.0


In [86]:
humidity= 3/5 * 0 +  2/5 * 0
humidity

0.0

In [89]:
humidity_tree=entropy([2./5.,3./5.], base=2)
humidity_tree

0.9709505944546688

In [90]:
gain_humidity=humidity_tree-humidity
print('Gain(humidity):', gain_humidity)

Gain(humidity): 0.9709505944546688


windy

In [3]:
false_info=entropy([1/3,2/3], base=2)
true_info=entropy([1/2,1/2], base=2)

print('Entropy (false):', false_info)
print('Entropy (true):', true_info)

Entropy (false): 0.9182958340544894
Entropy (true): 1.0


In [5]:
windy=3/5 * 0.9182958340544894 +  2/5 * 1
windy

0.9509775004326937

In [6]:
windy_tree=entropy([2./5.,3./5.], base=2)
windy_tree

0.9709505944546688

In [7]:
gain_windy=windy_tree-windy
print('Gain(windy):', gain_windy)

Gain(windy): 0.019973094021975113
